### DISCUSSION TITLE
json-prune within an array

### DISCUSSION THREAD

**Discussion by Procyon-b** [uBlockOrigin/uAssets#16710] [General]
I need to remove any element of an array based on the one of its child.
This works:
`##+js(json-prune, mailListElements.2 , mailListElements.2.rawData.promotext)`

but this doesn't:
`##+js(json-prune, mailListElements.[] , mailListElements.[].rawData.promotext)`

I don't want to do this:
`##+js(json-prune, mailListElements.[].* , mailListElements.[].rawData.promotext)`
because it leaves an empty object in that array element and the page's script doesn't like it.

Is it at all possible without using the array index?  Fortunately, ads in this array are placed in fixed position.

**Comment by krystian3w** [comment]
https://github.com/AdguardTeam/Scriptlets/issues/183 - maybe need follow these.

**Comment by farya977** [comment]
> I need to remove any element of an array based on the one of its child. This works: `##+js(json-prune, mailListElements.2 , mailListElements.2.rawData.promotext)`
> 
> but this doesn't: `##+js(json-prune, mailListElements.[] , mailListElements.[].rawData.promotext)`
> 
> I don't want to do this: `##+js(json-prune, mailListElements.[].* , mailListElements.[].rawData.promotext)` because it leaves an empty object in that array element and the page's script  doesn't like it.
> 
> Is it at all possible without using the array index? Fortunately, ads in this array are placed in fixed position.

It seems like you are trying to remove elements from an array based on the value of a child property, specifically the promotext property in the rawData object. While you've mentioned that the following syntax works: 
##+js(json-prune, mailListElements.2 , mailListElements.2.rawData.promotext)
You are facing issues when trying to remove elements using a wildcard approach. For instance, this doesn't work as expected:
##+js(json-prune, mailListElements.[] , mailListElements.[].rawData.promotext)
You also don't want to use the wildcard syntax with the .* placeholder because it leaves empty objects in the array elements, causing issues with the page's script.

Considering that the ads in the array are placed in fixed positions, you want a method to remove elements without using the array index.

Unfortunately, the JSONPath specification (which is likely used in your case) doesn't support a direct way to filter array elements based on their child property values without using array indices or wildcards. The wildcard approach with .* is the closest to achieving the desired result, but the empty objects left behind cause problems.

Since the JSONPath syntax doesn't provide an exact solution for your requirement, you may need to consider an alternative approach. One possible workaround is to perform this operation in two steps:

Use the wildcard approach to get the array indices that match the condition (rawData.promotext exists).
Then, remove those specific array elements using their indices explicitly.
Here's an example of how you can do it in JavaScript:
// Step 1: Find array indices with matching condition
let indicesToRemove = mailListElements.reduce((acc, element, index) => {
  if (element.rawData && element.rawData.promotext) {
    acc.push(index);
  }
  return acc;
}, []);

// Step 2: Remove elements from the array using indices
for (let i = indicesToRemove.length - 1; i >= 0; i--) {
  mailListElements.splice(indicesToRemove[i], 1);
}
This approach ensures that you remove the elements from the array without leaving empty objects behind. Just make sure to execute this code after the array mailListElements has been populated with the necessary data.
